lib/parsers.js

```diff
@@ -115,7 +115,7 @@ function getContextFromResponse (res, conf, isError) {
 
   if (isError) {
     context.headers_sent = res.headersSent
-    context.finished = res.finished
+    context.finished = Boolean(res.finished || res.writableEnded)
```

Comment:
We hit another test failure, this time in a different scenario. The failing tests shows this log

```
not ok 6 pseudo-res testCase 1
  ---
    operator: deepEqual
    expected: |-
      { status_code: 500, headers: {}, headers_sent: undefined, finished: undefined }
    actual: |-
      { status_code: 500, headers: {}, headers_sent: undefined, finished: false }
    at: <anonymous> (/home/runner/work/apm-agent-nodejs/apm-agent-nodejs/test/parsers.test.js:116:9)
```

There is an scenario we did not take in consideration. AOM agent can run not only on HTTP servers but also in other contexts like AWS Lambda. In such scenarios agent does not get an HttpResponse but an object with a similar shape. That object does not have `finished` or `writableEnded` therefore the `context.finished` shouldn't have a value so we need to drop this one liner for something with the following behaviour
- if both properties are `undefined` => context.finished = undefined
- if any of the properties is `boolean` => context.finished = the_boolean_value

I'll drop a suggestion but do not feel obligated to apply it directly, we can have a discussion :) 

```suggestion
    if (typeof res.finished === 'boolean') {
      context.finished = res.finished
    } else {
      context.finished = res.writableEnded
    }
```

Reply:
Hey @david-luna, please tell me if I understand it correctly. 
1. If `res.finished` is of type boolean then either `res.writableEnded` will be the same boolean value or `undefined`. So anyways it will evaluate to `res.finished`. 
2. I am assuming that if `res.finished` is `undefined` then `res.writableEnded` will also be `undefined` by default because it will not be an HTTP server then. Is this assumption correct?
3. If we come to "else" statement, then we have `res.finished` as `undefined` and this implies from 2nd statement that `res.writableEnded` is also `undefined`. So can't we directly evaluate the expression to `undefined` over here. 


And hence the compiling above three points, I am thinking something like this:
```suggestion
    if (typeof res.finished === 'boolean') {
      context.finished = Boolean(res.finished || res.writableEnded)
    } else {
      context.finished = undefined
    }
    
```

Reply:
Hi @thesixpathguy 

sorry for the delay on the feedback. I was off for a week

> If `res.finished` is of type boolean then either `res.writableEnded` will be the same boolean value or `undefined`. So anyways it will evaluate to `res.finished`.

So if `finished === false` and `writableEnded === undefined` we got the expression `false || undefined` resulting in `undefined` and this is not what we need. Using the `Boolean` constructor solves this but if we already know that `res.finished` contains a valid value we could use it directly.

> I am assuming that if `res.finished` is `undefined` then `res.writableEnded` will also be `undefined` by default because it will not be an HTTP server then. Is this assumption correct?

At this moment your assumption is good but a future node release will remove `finished` and `writableEnded` will be the only property available in the HTTP response object. The issue is about covering this possible scenario

The code I've suggested before was hiding some context by merging the new scenario (only writableEnded defined) with the scenario of non HTTP Responses (lambda instrumentation). We can use some extra keystrokes to be even more explicit

```javascript
if (typeof res.finished === 'boolean') {
    // Scenario: versions of node with finished defined, we can still use it
    context.finished = res.finished
} else if (typeof res.writableEnded === 'boolean') {
    // Scenario: new versions of node with finished removed, we use writableEnded
    context.finished = res.writableEnded
}
// If no branch was executed there is no need to set finished property for context
// because it was a response like object from lambda instrumentation
```

Reply:
@david-luna 
Oh, I see. 
Yup, the above code seems correct and concise. Let me implement that one.